include/dd/RealNumberUniqueTable.hpp

```diff
@@ -0,0 +1,209 @@
+#pragma once
+
+#include "dd/Definitions.hpp"
+#include "dd/MemoryManager.hpp"
+#include "dd/UniqueTableStatistics.hpp"
+
+#include <array>
+#include <cstddef>
+#include <cstdint>
+#include <iostream>
+
+namespace dd {
+
+struct RealNumber;
+
+/**
+ * @brief A unique table for real numbers.
+ * @details A hash table that stores real numbers. The hash table is implemented
+ * as an array of buckets, each of which is a linked list of entries. The hash
+ * table has a fixed number of buckets.
+ * @note: The implementation assumes that all values are non-negative and in the
+ * range [0, 1]. While numbers outside of this range can be stored, they will
+ * always be placed in the same bucket and will therefore cause collisions.
+ */
+class RealNumberUniqueTable {
+  /**
+   * @brief The number of buckets in the table.
+   * @details The number of buckets is fixed and cannot be changed after the
+   * table has been created. Increasing the number of buckets reduces the
+   * number of collisions, but increases the memory usage.
+   * @attention The number of buckets has to be one larger than a power of two.
+   * Otherwise, the hash function will not work correctly.
+   */
+  static constexpr std::size_t NBUCKET = 65537U;
+
+  /**
+   * @brief The initial garbage collection limit.
+   * @details The initial garbage collection limit is the number of entries that
+   * must be present in the table before garbage collection is triggered.
+   * Increasing this number reduces the number of garbage collections, but
+   * increases the memory usage.
+   */
+  static constexpr std::size_t INITIAL_GC_LIMIT = 65536U;
+
+public:
+  /**
+   * @brief The default constructor
+   * @param manager The memory manager to use for allocating new numbers.
+   * @param initialGCLim The initial garbage collection limit.
+   */
+  explicit RealNumberUniqueTable(MemoryManager<RealNumber>& manager,
+                                 std::size_t initialGCLim = INITIAL_GC_LIMIT);
+
+  /**
+   * @brief The hash function for the hash table.
+   * @details The hash function for the table is a simple linear (clipped) hash
+   * function. The hash function is used to map floating point numbers to the
+   * buckets of the table.
+   * @param val The floating point number to hash. Must be non-negative.
+   * @returns The hash value of the floating point number.
+   */
+  static std::int64_t hash(fp val) noexcept;
```

Comment:
At first, I was totally on board with your conclusion. Then I remembered that I actually implemented it like that back in the days because the signed version leads to more efficient assembly (especially when using gcc).
See https://godbolt.org/z/hbEWh4vsh.
So I'd rather keep it the way it is.

Reply:
Wow, that is a good catch! Would you mind adding this argument to the comments?

Reply:
Done in https://github.com/cda-tum/mqt-core/pull/373/commits/f28fb26086d0b49b6fedb5365f31c5ba2ce7d34d.
I hope the Godbolt link stays active for a while. Fortunately it's quite easy to replicate if that should happen.